feat(client): v3: ilp over http

sklarsa opened the pull request:
Adds a new client version to support the ILP protocol over HTTP transport, as well as the new config string syntax.

puzpuzpuz commented on sender.go:
```diff
@@ -61,25 +64,47 @@ const (
 //
 // Each sender corresponds to a single TCP connection. A sender
 // should not be called concurrently by multiple goroutines.
 type LineSender struct {
```
Keeping this struct as 1-in-2 makes it harder to maintain. TCP and HTTP senders have different behavior, e.g. around retry, so it's better to split them. I suggest introducing `HttpLineSender` struct for ILP/HTTP and exposing the following constructor functions: `NewLineSender`, `NewLineSenderFromConf`, `NewHttpLineSender`, `NewHttpLineSenderFromConf`. This way we won't break existing client code and keep implementations separate. Also, buffer related stuff could be extracted into a common struct which would be embedded into both `LineSender` and `HttpLineSender` - this way we'll avoid code duplication.

sklarsa replied:
> NewHttpLineSenderFromConf

I don't think we need this since we specify the protocol in the conf string itself. Maybe `NewLineSenderFromConf` is enough? And possible alias `NewLineSender` to `NewTcpLineSender` for clarity? (while keeping the existing function so as not to break compatibility)

puzpuzpuz replied:
It won't be possible to return one of two different structs from a single function without an interface. That's why we'd need different FromConf functions.

sklarsa commented on sender.go:
```diff
@@ -144,6 +169,63 @@ func WithFileNameLimit(limit int) LineSenderOption {
 	}
 }
 
+func WithGraceTimeout(timeout time.Duration) LineSenderOption {
```
Ok cool! I was going off the document (https://docs.google.com/document/d/1EQZjYrJO_Ll8RrSIZpbr15Ul3N-xAtTxvrunI-TpOR8/edit#heading=h.lrrfjiq35q9t) but I guess things have changed since then? HTTP timeout makes much more sense to me

puzpuzpuz replied:
> but I guess things have changed since then?

Probably. @jerrinot do you happen to know why Java HTTP sender has HTTP timeout, but not those grace settings?

jerrinot replied:
I discussed this with @amunra and I'm going to add this to Java client too

puzpuzpuz replied:
OK, we'll have to keep using these new settings then. Resolving this one.

sklarsa commented:
@puzpuzpuz what do you think about this approach? Before I go any further with this...

I've broken out `sender.go` into several files:
1. buffer.go for buffer-related things and the public `Write...` methods
2. tcp.go for `LineSender` Flush/constructor implementation
3. http.go for `HttpLineSender` Flush/constructor implementation

I've embedded `buffer` into both `LineSender` and `HttpLineSender` with no issues, except for moving the auto-flush on buffer size exceeding the max into each client separately (instead of the end of the `at` method).  I'm not super happy with the code duplication in the public methods of `LineSender` and `HttpLineSender` (mostly the godocs), but there's no other way to create method chains with the `Write...` funcs since we're not using interfaces.  There will also be duplication in some of the options constructors for options that are shared between tcp and http (like tls or init buffer size).

The next steps will be related to config parsing. I should be able to reuse most of my code, but as you mentioned earlier today, will need separate methods for https and tcp, so I'll probably return errors for invalid config settings (like http timeout for tcp etc)

What do you think of this approach?

puzpuzpuz commented on go.mod:
```diff
@@ -1,4 +1,4 @@
+module github.com/questdb/go-questdb-client/v3
 
 go 1.17
 
```
Go 1.17 is quite old, so it's not a big deal if we upgrade minimal Go version. It'll be much easier to do that now since we're going to release a major version.

puzpuzpuz commented:
> What do you think of this approach?

I like the approach, thanks! Code duplication won't be huge, so it's not a big deal IMO. Keeping implementations separate will make the code more readable and maintainable which is much more important.

sklarsa commented on sender.go:
```diff
@@ -156,95 +238,153 @@ func NewLineSender(ctx context.Context, opts ...LineSenderOption) (*LineSender,
 
 // Close closes the underlying TCP connection. Does not flush
 // in-flight messages, so make sure to call Flush first.
 func (s *LineSender) Close() error {
```
I went with your suggestion with the global transport

sklarsa commented:
> > What do you think of this approach?
> 
> I like the approach, thanks! Code duplication won't be huge, so it's not a big deal IMO. Keeping implementations separate will make the code more readable and maintainable which is much more important.

Ok, so I've done that and _more_...

This afternoon, I started to realize that it's awkward to have `LineSender` and `HttpLineSender` in the same package. For example, if I wanted to add a custom address to my sender, I need to use `WithAddress` for `LineSender` and `WithHttpAddress` for `HttpLineSender`, and would get an annoying compilation error if I use the wrong combination.  

So I split everything out into 3 separate packages, `pkg/buffer`, `pkg/tcp` and `pkg/http`. This required me to change many of the buffer's fields and methods to public, but I don't think that should be a major problem. Now, the LineSender API and usage is the same across both tcp and http.  To solve the compatibility issue, I aliased all of the funcs and structs from the `tcp` package in the root package `v3`. 

If you think that this approach is solid, I can work on writing more tests. I reverted many of my older tests that used the `LineSender` for both, just to get the thing to build. I can also work on cleanup stuff like godocs, and eventually auto flush for http.

puzpuzpuz commented:
> If you think that this approach is solid, I can work on writing more tests.

Yes, the approach looks good to me. Using separate packages makes sense.

sklarsa commented on sender.go:
```diff
@@ -156,95 +238,153 @@ func NewLineSender(ctx context.Context, opts ...LineSenderOption) (*LineSender,
+		tr := &http.Transport{
+			Proxy: http.ProxyFromEnvironment,
+			DialContext: defaultTransportDialContext(&net.Dialer{
+				Timeout:   30 * time.Second,
```
I removed this DialContext and instead am going with per-request timeouts as discussed earlier

sklarsa commented:
ok @puzpuzpuz, should be good for re-review

puzpuzpuz commented on README.md:
```diff
@@ -10,7 +10,7 @@ Features:
-Documentation is available [here](https://pkg.go.dev/github.com/questdb/go-questdb-client/v2).
+Documentation is available [here](https://pkg.go.dev/github.com/questdb/go-questdb-client/v3).
 
 ## Usage
```
An example with HTTP sender is still missing. It's important to mention that the client supports both ILP/TCP and ILP/HTTP and have an example in the readme. WDYT?

puzpuzpuz commented on examples.manifest.yaml:
```diff
@@ -20,13 +20,43 @@
+  header: |-
+    Go client library [docs](https://pkg.go.dev/github.com/questdb/go-questdb-client/v3)
+    and [repo](https://github.com/questdb/go-questdb-client).
+  #auth:
```
Looks like a leftover.

sklarsa replied:
I wasn't sure what to do with those fields, since I don't know the mechanics of how this is translated into the autodocs for questdb.io and the cloud samples. My http auth examples use different fields (user/pass & token), and I'm not sure what modifications I need to make to any other libs for this to work

puzpuzpuz replied:
This yaml is used in Cloud UI. Currently, we only have ILP/TCP snippets there, so the `auth` section is specific to ILP/TCP.

@insmac @amunra @jerrinot we should add ILP/HTTP snippets, at least for the clients that support it. This patch adds some examples for the Go client, so we should probably do the same in Java/C/Python/Rust clients and then update Cloud UI. WDYT?

puzpuzpuz commented on examples.manifest.yaml:
```diff
@@ -20,13 +20,43 @@
   addr:
     host: localhost
     port: 9009
+- name: http
```
nit: I'd add `ilp-` suffix to the example names: `ilp-http`, `ilp-http-auth`, `ilp-http-auth-tls`

puzpuzpuz commented on examples.manifest.yaml:
```diff
@@ -20,13 +20,43 @@
+  header: |-
+    Go client library [docs](https://pkg.go.dev/github.com/questdb/go-questdb-client/v3)
+    and [repo](https://github.com/questdb/go-questdb-client).
+  #auth:
```
nit: another leftover here.

sklarsa replied:
same as above

puzpuzpuz commented on examples/http/auth-and-tls/main.go:
```diff
@@ -0,0 +1,62 @@
+	"log"
+	"time"
+
+	qdb "github.com/questdb/go-questdb-client/v3/pkg/http"
```
nit: do we need the `/pkg` section in the package path? Why not `github.com/questdb/go-questdb-client/v3/http`?

sklarsa replied:
No reason, just trying to follow go conventions. I can move `http` and `tcp` out of `pkg/`

puzpuzpuz replied:
> No reason, just trying to follow go conventions.

Do you refer to https://go.dev/blog/package-names (couldn't find /pkg convention there) or it's something else?

sklarsa replied:
I was using the golang standards repo (https://github.com/golang-standards/project-layout/tree/master/pkg) where `pkg` is used for library code that can be imported by external projects

puzpuzpuz replied:
I see. That's not official convention. Interestingly, it simply follows an old convention from Golang source code:
> The pkg directory origins: The old Go source code used to use pkg for its packages and then various Go projects in the community started copying the pattern

sklarsa replied:
Most go projects I've seen use `pkg` these days. Interesting to learn that it's not an official convention

puzpuzpuz commented on const.go:
```diff
@@ -0,0 +1 @@
+package questdb
```
Do we need this file?

puzpuzpuz commented:
> I'm not sure what is the best way to generalize the integration tests to reuse the test cases across tcp & http. Do you have any ideas?

We'd need to extract the set up/tear down parts (start/stop QDB + proxy). As for the sender - should be enough to define an interface in the tests and use it to interact with the sender.

puzpuzpuz commented on pkg/buffer/buffer.go:
```diff
@@ -0,0 +1,640 @@
+	return b
+}
+
+func (b *Buffer) At(ts time.Time, sendTs bool) error {
```
nit: would be nice to document this method since it's public.

puzpuzpuz commented on pkg/http/sender.go:
```diff
@@ -0,0 +1,605 @@
+	globalTransport http.Transport = http.Transport{
+		Proxy:                 http.ProxyFromEnvironment,
+		MaxConnsPerHost:       0,
+		MaxIdleConns:          100,
```
We should expose this as a sender setting. To achieve that, we could create the transport lazily, i.e. on the first sender initialization.

sklarsa replied:
What about making a `WithTransport` option for users to customize their sender's transport? We would default to the global transport.  

This would also be a nice workaround for the global WithTlsInsecureSkipVerify setting problem (once it's set, it can't be unset because we only have a single tls config for the globalTransport). If you want multiple senders with different tls configs, you can do that with a custom transport option.  Wdyt?

puzpuzpuz commented on pkg/http/sender.go:
```diff
@@ -0,0 +1,605 @@
+		MaxConnsPerHost:       0,
+		MaxIdleConns:          100,
+		MaxIdleConnsPerHost:   100,
+		IdleConnTimeout:       90 * time.Second,
```
Same here and for the below settings: we need to let users configure them.

sklarsa replied:
I had problems writing integration tests with the global transport since there was a dependency on test ordering (to ensure that `TlsMode` was not used after `TlsInsecureVerify` was).  So I added a transport override option in [this commit](https://github.com/questdb/go-questdb-client/pull/26/commits/7b2263b4ba581ce55a71bc694c0f1fed4534123f). WDYT?

puzpuzpuz replied:
Looks like a flexible enough option for folks who want to tweak the transport. 👍

puzpuzpuz commented on pkg/http/sender.go:
```diff
@@ -0,0 +1,605 @@
+}
+
+// WithBasicAuth sets a Basic authentication header for
+// ILP requests over HTTP
```
nit:
```suggestion
// ILP requests over HTTP.
```

puzpuzpuz commented on pkg/http/sender.go:
```diff
@@ -0,0 +1,605 @@
+}
+
+// WithBearerToken sets a Bearer token Authentication header for
+// ILP requests
```
nit:
```suggestion
// ILP requests.
```

puzpuzpuz commented on pkg/http/sender.go:
```diff
@@ -0,0 +1,605 @@
+}
+
+// WithGraceTimeout is used in combination with min throughput
+// to set the timeout of an ILP request. Defaults to 5 seconds
```
nit:
```suggestion
// to set the timeout of an ILP request. Defaults to 5 seconds.
```

puzpuzpuz commented on pkg/http/sender.go:
```diff
@@ -0,0 +1,605 @@
+}
+
+// WithMinThroughput is used in combination with grace timeout
+// to set the timeout of an ILP request. Defaults to 100KiB/s
```
nit:
```suggestion
// to set the timeout of an ILP request. Defaults to 100KiB/s.
```

puzpuzpuz commented on pkg/http/sender.go:
```diff
@@ -0,0 +1,605 @@
+			return nil
+		}
+		// Otherwise, we will retry sending the request until the retry timeout is breached
+		err = fmt.Errorf("Non-OK Status Code %d: %s", resp.StatusCode, resp.Status)
```
We should parse error messages sent from the server and propagate them to the user. See `JsonErrorParser` in the Java client and its usage:
https://github.com/questdb/questdb/blob/713847e72f0c6977b61bf3001094dc60c6939c74/core/src/main/java/io/questdb/cutlass/line/http/LineHttpSender.java#L512

puzpuzpuz commented on pkg/http/sender.go:
```diff
@@ -0,0 +1,605 @@
+	if err != nil {
+		return err
+	}
+	if s.Len() > s.BufCap {
```
Other ILP/HTTP clients don't have this logic. It can be met only in ILP/TCP clients. For HTTP we flush either on `Flush` or when `auto_flush_rows` is reached. The same applies to `At`.

sklarsa replied:
What is the reason for that? I can easily remove the code (and the `WithBufferCapacity` http sender option too), I'm just curious

puzpuzpuz replied:
That's because in ILP/HTTP we want the user to have full control over transaction boundaries, i.e. by default commit(s) should only happen on `Flush`. Then the `auto_flush_rows` setting is exposed for those users who don't care about transactions and just let the client flush occasionally.

puzpuzpuz commented on pkg/tcp/sender.go:
```diff
@@ -237,8 +243,90 @@ func NewLineSender(ctx context.Context, opts ...LineSenderOption) (*LineSender,
+	return opts, nil
+}
+
+// LineSenderFromConf creates a LineSender using the QuestDB config string format.
```
Would be great to have a few config string examples here.

puzpuzpuz commented on pkg/http/sender.go:
```diff
@@ -0,0 +1,605 @@
+	return s, nil
+}
+
+// LineSenderFromConf creates a LineSender using the QuestDB config string format.
```
Would be great to have a few config string examples here.

puzpuzpuz commented on tests/interop/test/interop/ilp-client-interop-test.json:
```diff
@@ -0,0 +1,1634 @@
+[
```
We should not check-in this file directly, but use a git submodule instead. That's what we do in all client repos.

puzpuzpuz commented on tests/integration/sender_integration_test.go:
```diff
@@ -97,19 +111,27 @@ func setupQuestDB0(ctx context.Context, auth ilpAuthType, setupProxy bool) (*que
 		"QDB_CAIRO_MAX_UNCOMMITTED_ROWS":        "1",
 		"QDB_LINE_TCP_MAINTENANCE_JOB_INTERVAL": "100",
-		"QDB_PG_ENABLED":                        "false",
+		"QDB_PG_ENABLED":                        "true",
```
nit: do we use PGWire in tests?

sklarsa replied:
We don't directly, but the container healthchecks fail when `QDB_PG_ENABLED=false`:
```
2024-03-05T19:52:40.664738Z I i.q.c.h.p.StaticContentProcessor [96] incoming [url=/]
2024-03-05T19:52:41.664923Z I http-server scheduling disconnect [fd=96, reason=12]
2024-03-05T19:52:41.664952Z I http-server disconnected [ip=10.89.3.36, fd=96, src=queue]
```

puzpuzpuz replied:
Hmm, that's unexpected. Could you check with a standalone Docker container? Maybe we have a bug on the server side.

puzpuzpuz reviewed (commented):
The PR looks solid! A few more changes are necessary to finalize it: auto_flush support and flush logic in the HTTP sender.

amunra commented:
The config options have been changing over time (sorry).

E.g. `grace_timeout` has now been renamed to the more sensible `request_timeout` and there's also a new `auto_flush_interval`.

The doc has been updated.

I suspect we need a cross-review of all the clients to check that any differences are intentional.

In the past we've also had some cross-client tests. I don't know if there's any way if which we can introduce something similar again.

puzpuzpuz commented:
The tests seem to be failing.

puzpuzpuz commented on README.md:
````diff
@@ -59,3 +64,51 @@ func main() {
 ```
+
+### HTTP
+
````
nit: would be great a doc link somewhere here. This would be helpful for users who don't know that we have ILP/TCP and ILP/HTTP and Go client supports both of them.
https://questdb.io/docs/reference/api/ilp/overview/#http-and-tcp-overview

sklarsa commented on conf.go:
```diff
@@ -0,0 +1,216 @@
+				}
+			case "auto_flush":
+				if value.String() == "on" {
+					return opts, NewConfigStrParseError("auto_flush option is not supported")
```
sounds good! Added for both tcp & http LineSenders

puzpuzpuz replied:
I don't think we need auto flush in the ILP/TCP client. See https://github.com/questdb/go-questdb-client/pull/26#discussion_r1515695679

puzpuzpuz commented on pkg/tcp/sender.go:
```diff
@@ -62,24 +55,26 @@ const (
+	key     string // Erased once auth is done.
+
+	// Auto-flush fields
+	autoFlush     bool
```
I don't think we have auto_flush behavior in any other TCP client, e.g. Java's ILP/TCP sender doesn't have it. ILP/TCP senders are kept as is when it comes to flushing/buffer management.

@jerrinot please correct me if I'm wrong.

sklarsa replied:
removed from tcp sender

puzpuzpuz commented on pkg/http/sender.go:
```diff
@@ -0,0 +1,638 @@
+
+// WithAutoFlushRows sets the number of buffered rows that
+// must be breached in order to trigger an auto-flush.
+// Defaults to 75000.
```
Looks like in Java this setting defaults to 600. Worth double-checking other clients and having a common default.

sklarsa replied:
> Looks like in Java this setting defaults to 600. Worth double-checking other clients and having a common default.

@amunra the auto-flush rows default in the ILP Client Configuration document is 75k. Should I use 75k or 600 here?

puzpuzpuz commented on pkg/http/sender.go:
```diff
@@ -0,0 +1,638 @@
+	if s.tlsMode > 0 {
+		uri += "s"
+	}
+	uri += fmt.Sprintf("://%s/write", s.address)
```
nit: uri doesn't change during the sender lifecycle, so we could allocate it a single time, on sender initialization.

puzpuzpuz commented on pkg/http/sender.go:
```diff
@@ -0,0 +1,638 @@
+	requestTimeout              time.Duration
+
+	// Auto-flush fields
+	autoFlush     bool
```
We don't need this flag in the sender. `autoFlushRows == 0` could mean that auto flush is disabled.

puzpuzpuz commented on pkg/http/sender.go:
```diff
@@ -0,0 +1,638 @@
+// Close closes the underlying HTTP client. If no clients remain open,
+// the global http.Transport will close all idle connections.
+func (s *LineSender) Close() {
+	newCt := clientCt.Add(-1)
```
We should check auto flush here and, if it's enabled, flush the sender here.

Also, we should have a `closed` flag in the sender, to avoid weird behavior (multiple `newCt` decrement) on double close.

sklarsa replied:
I added a context to the Close argument list to pass to `s.Flush` in the body of the function. I updated all tests and also added error checking to the `defer` calls in the examples

puzpuzpuz commented on pkg/http/sender.go:
```diff
@@ -0,0 +1,638 @@
+
+	s.msgCount++
+
+	if s.autoFlush && s.msgCount > s.autoFlushRows {
```
Should be enough to have `s.msgCount == s.autoFlushRows` as the condition here. This way we'll flush when we have exactly `autoFlushRows` rows sitting in the buffer.

sklarsa commented:
> > I'm not sure what is the best way to generalize the integration tests to reuse the test cases across tcp & http. Do you have any ideas?
> 
> We'd need to extract the set up/tear down parts (start/stop QDB + proxy). As for the sender - should be enough to define an interface in the tests and use it to interact with the sender.

I played around with the interface idea, and it actually won't work due to the return types of the `Float64Column()`, `BoolColumn()` etc methods. If I make an interface with a method that returns an instance of that interface, all structs that implement the interface must also return the instance of the interface, not a "subclass" of it. So it's difficult to introduce the interface in tests only without modifying the actual package code.

I ran some benchmarks and was able to see the interface penalty, so I guess I'll just figure out another way to generalize the relevant tests. Maybe using a common config that can be used to generate both tcp/http test fixtures.

puzpuzpuz commented on pkg/http/sender.go:
```diff
@@ -0,0 +1,660 @@
+	}
+
+	if s.autoFlushRows > 0 && s.Buffer.Len() > 0 {
+		err := s.Flush(ctx)
```
nit: this logic should be reflected in the method's godoc.

sklarsa commented on pkg/http/sender.go:
```diff
@@ -0,0 +1,660 @@
+		reqCtx,
+		http.MethodPost,
+		s.uri,
+		s,
```
The buffer is set in the `Buffer.WriteTo` method that's called by the http client.  I added a test confirming this behavior

puzpuzpuz replied:
I see. It looks like an internal standard library's fast path since they haven't documented it properly:
https://github.com/golang/go/blob/a46ecdca36b2359e7496c6d694703a9a9706d760/src/net/http/client.go#L706
https://github.com/golang/go/blob/a46ecdca36b2359e7496c6d694703a9a9706d760/src/io/io.go#L408-L412

If std lib stops checking `WriterTo` interface in future, our sender will break. Worth at least adding a clarifying comment.

sklarsa commented on pkg/http/sender.go:
```diff
@@ -0,0 +1,660 @@
+			}
+		}
+	}
+
```
I moved the msgCount implementation to the buffer so we can update it with lastMsgPos.  But I had a thought, what do we do with message count when `buffer.WriteTo()` returns an error and leaves a partially-written buffer? We correctly decrement lastMsgPos, but there's no way to tell how many messages were successfully written.

puzpuzpuz replied:
In that case, we could reset sender's state since such an error means that the connection got broken. In case of TCP, the sender should be closed by user. In case of HTTP, the sender may be reused to write messages once again and try sending them.

puzpuzpuz commented on pkg/http/errors.go:
```diff
@@ -0,0 +1,47 @@
+
+	Code    string `json:"code"`
+	Message string `json:"message"`
+	Line    int    `json:"line"`
```
The "line" field is optional:
```suggestion
	Line    int    `json:"line,omitempty"`
```

puzpuzpuz commented on pkg/tcp/sender.go:
```diff
@@ -62,24 +55,23 @@ const (
-	hasTable      bool
-	hasTags       bool
-	hasFields     bool
+	buf *buffer.Buffer
```
nit: could be `buffer.Buffer`

sklarsa commented on conf_parse.go:
```diff
@@ -0,0 +1,243 @@
+		switch strings.ToLower(k) {
+		case "addr":
+			senderConf.address = v
+		case "user":
```
@puzpuzpuz @amunra  is it "user" or "username"?

puzpuzpuz replied:
It's "user"

puzpuzpuz replied:
It's "user"

sklarsa replied:
The python docs have "username"

```
from questdb.ingress import Sender, ServerTimestamp

conf = (
    'tcp::addr=localhost:9009;' +
    'username=testUser1;' +
    'token=5UjEMuA0Pj5pjK8a-fa24dyIf-Es5mYny3oE_Wmus48;' +

...
```

puzpuzpuz replied:
Hmm, same with Rust client. Let's change it to username in Go client too.

cc @jerrinot since Java sender expects "user", not "username".

puzpuzpuz replied:
Updated in 40f560a4233821f7f2b5377650e948b52ce3986c

sklarsa commented on sender.go:
```diff
@@ -25,32 +25,146 @@
 
 const (
-	defaultBufferCapacity = 128 * 1024
+	defaultHttpAddress = "127.0.0.1:9009"
```
@puzpuzpuz http is still on port 9000, right?

puzpuzpuz replied:
Yes, it's a leftover. Nice find. Fixed in dc95225f403d9b35a94425b749c634d397fedcf7

sklarsa commented:
@puzpuzpuz I'm comfortable with it.. all tests (including integration) are passing and coverage is very high
